### PR TITLE
OAK-10803 - Modified DocumentPropertyState.java to compress values ba…

### PR DESCRIPTION
**Pull request by ionutzpi**
Added compression for values in DocumentPropertyState

### CODE REVIEW

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: I would keep this private.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: If there is an exception while compressing, we should return the uncompressed value rather than throwing the exception.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: please log the exception.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: please log the exception.

**Review by rishabhdaim — CHANGES_REQUESTED** [review]

**Comment by rishabhdaim** [comment]
@ionutzpi could you please add test cases for this change.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (stefan-egli) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: naming .. but maybe it could indicate what it is used for, eg `TEST_NODE` ?

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (stefan-egli) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: would org.apache.commons.lang3.RandomStringUtils (maybe using the variant that accepts a `Random` that can be seeded with a fixed value) be a useful alternative for this hard coded string?

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (stefan-egli) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: minor: could also leave this method unchanged as it is unrelated and only cosmetic

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: I wouldn't log the value as that could leak sensitive data into the log file. While we don't have the path, perhaps just log the property name plus the IOException message?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: an IOException on decompress is bad, as that would fail the reading of the value (while the same on compression is not a problem, then it would just not compress). So I'd argue this would justify a `LOG.error` here.

problem might then be, that in IOException case this might become a noisy logger .. not sure if we should already consider that now though...

but also : I wouldn't log the value itself.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: could/should this be `final` as well? (just one more `null` assignment necessary..)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]:
```suggestion
            compressedValue = compress(value.getBytes());
```
formatting

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: with now returning the original uncompressed `value` here in exception case, that then likely leads to subsequent issues with uncompressing. It might need more complex logic in exception cases I think.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: we could replace this  `"<any string>".repeat(<count>)`.

Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: this could be made private as well.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: please log a `WARN` message here.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java:215]: please revert these un-necessary changes.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java:226]: same as above.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: I would suggest to create a test case with compression/decompression throwing exceptions

**Comment by Joscorbe** [comment]
I'm fine with the tests, but both `stringBelowThresholdSize()` and `stringAboveThresholdSize()` are essentially the same - which is correct from a black box test point of view.

However, I wonder if you could verify (with mockito for example) that the compressing/uncompressing was actually done in one of the tests and not in the other. Otherwise I feel these tests would pass even without using the new implemented code at all.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (Joscorbe) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java:133]: I'm fine with these tests, but both stringBelowThresholdSize() and stringAboveThresholdSize() are essentially the same - which is correct from a black box test point of view.

However, I wonder if you could verify (with mockito for example) that the compressing/uncompressing was actually done in one of the tests and not in the other. Otherwise I feel these tests would pass even without using the new implemented code at all.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Could you please remove the format changes.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Same as above.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: Please avoid `*` imports and import only required packages.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: This could probably be package protected?

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (stefan-egli) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: could you wrap this in a `assertEquals()` or similar?

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java:133]: +1 to assert that compression actually happened (which is indirectly already done by asserting null/non-null for compressedValue, but it could in addition check that it did some compression indeed and not leave the value unchanged - which btw could also be a nice test case, to pass a compression type that doesn't actually compress - it should then still also work)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (reschke) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: 1. Can we avoid creating a stream wrapper here, even if this needs changes in compression?
2. Constructing a string from bytes should be done with an explicit character encoding, otherwise it's platform/locale dependant.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (reschke) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: value.getBytes() should specify the character encoding to use (note UTF-8 is not ncessarily the default).

**Review by reschke — CHANGES_REQUESTED** [review]

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (reschke) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: private maybe?

Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: No, I need it in other class.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Done

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: Done

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Done point 2. For 1. I wouldn't add changes in compression interface, only if indeed required.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: Done

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Done

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Done

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java:133]: Done

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: the revert for this wasn't done though?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: the revert for this wasn't done though?

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: thx
(nitpick: I think then you could remove this line as they getValue is now called twice)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: The revert was done.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: The revert was done.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: Done.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Done. I reverted readArrayProperty doc changes. Thanks.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Done. I reverted also readArrayProperty doc changes. As per Stefan suggestion, thanks

**Review by stefan-egli — APPROVED** [review]
+1, lgtm. I see @Joscorbe  and @reschke have to yet comment on their concerns.

**Comment by ionutzpi** [comment]
Hi @Joscorbe  and @reschke could you please check the PR and approve it if is fine?

**Comment by Joscorbe** [comment]
+1 from me.

**Review by Joscorbe — APPROVED** [review]

**Comment by ionutzpi** [comment]
Hi @reschke , @Joscorbe, @stefan-egli could we close this PR? @reschke it need only your approval. Thanks

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (reschke) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: no wildcard imports, please

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (reschke) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Just in the test class? Can we avoid exposing it just for that purpose?

**Review by reschke — CHANGES_REQUESTED** [review]

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java
Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java]: Done

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java:171]: I think that the access modifier is fine as default for this method.

Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Done. I made the modifier for the method private.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]:
```suggestion
        String localValue = value;
        String localCompressedValue = null;
        if (compression != null && size > DEFAULT_COMPRESSION_THRESHOLD ) {
            try {
                localCompressedValue = compress(value.getBytes(StandardCharsets.UTF_8));
                localValue = null;
            } catch (IOException e) {
                LOG.warn("Failed to compress property {} value: ", name, e);
            }
        }
        this.value = localValue;
        this.compressedValue = localCompressedValue;
```

A last-minute idea : what about making `value` and `compressedValue` `final` again and assign them like shown above? Having them final would make it somewhat stronger and clearer in my view. Wdyt?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Two more things came to mind:
* we should have a way to disable compression - eg by making this value `-1`. Currently that is not the case.
* are we sure we want the default to be 1024? We could choose -1 instead until we are confident this change should be on by default.

**Review by stefan-egli — COMMENTED** [review]
Two more last minute comments added

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java]: Done. I set default to -1

**Comment by reschke** [comment]
FWIW, you don't *need* my approval. What you need is an Oak committer willing to merge this on your behalf.

I didn't have time to review this properly yet.

**Review by stefan-egli — APPROVED** [review]
+1, PR looks good to me now, thx for the changes

**Pull request merged**
